frontend/test/resolution/testTuples.cpp

```diff
@@ -846,6 +845,262 @@ static void testTupleGeneric() {
+  assert(m["z"].kind() == QualifiedType::VAR);
+  assert(m["z"].type()->isIntType());
+}
+
+static void test27() {
+  printf("%s\n", __FUNCTION__);
+  Context ctx;
+  auto context = &ctx;
+  ErrorGuard guard(context);
+
+  std::string program =
+    R""""(
+    var g: (int, int);
+    iter myIter() const ref: (int, int) { yield g; }
+
+    for (i, j) in myIter() {
+      var z = i;
+    }
+    )"""";
+
+  // The entire tuple (i, j) is itself a 'REF' tuple.
+  auto mod = parseModule(context, program);
+  auto& rr = resolveModule(context, mod->id());
+  auto astTup = mod->stmt(2)->toIndexableLoop()->index();
+  assert(astTup->isTupleDecl());
+  auto& qtTup = rr.byAst(astTup).type();
+  assert(qtTup.kind() == QualifiedType::CONST_REF);
+  assert(qtTup.type()->isTupleType());
+
+  // The de-tupled components still maintain their 'REF'ness.
+  auto m = resolveTypesOfVariables(context, program, { "i", "j", "z" });
+  assert(!guard.realizeErrors());
+  assert(m["i"].kind() == QualifiedType::CONST_REF);
+  assert(m["i"].type()->isIntType());
+  assert(m["j"].kind() == QualifiedType::CONST_REF);
+  assert(m["j"].type()->isIntType());
+  assert(m["z"].kind() == QualifiedType::VAR);
+  assert(m["z"].type()->isIntType());
+}
+
+static void test28() {
+  printf("%s\n", __FUNCTION__);
+  Context ctx;
+  auto context = &ctx;
+  ErrorGuard guard(context);
+
+  std::string program =
+    R""""(
+    var g: (int, int);
+    ref (a, b) = g;
+    )"""";
+
+  // The entire tuple (i, j) is itself a 'REF' tuple.
+  auto mod = parseModule(context, program);
+  auto& rr = resolveModule(context, mod->id());
+  auto astTup = mod->stmt(1);
+  assert(astTup->isTupleDecl());
+  auto& qtTup = rr.byAst(astTup).type();
+  assert(qtTup.kind() == QualifiedType::REF);
+  assert(qtTup.type()->isTupleType());
+
+  // The de-tupled components still maintain their 'REF'ness.
+  auto m = resolveTypesOfVariables(context, program, { "a", "b" });
+  assert(!guard.realizeErrors());
+  assert(m["a"].kind() == QualifiedType::REF);
+  assert(m["a"].type()->isIntType());
+  assert(m["b"].kind() == QualifiedType::REF);
+  assert(m["b"].type()->isIntType());
+}
+
+static void test29() {
+  printf("%s\n", __FUNCTION__);
+  Context ctx;
+  auto context = &ctx;
+  ErrorGuard guard(context);
+
+  std::string program =
+    R""""(
+    var g: (int, int);
+    const ref (a, b) = g;
+    )"""";
+
+  auto mod = parseModule(context, program);
+  auto& rr = resolveModule(context, mod->id());
+  auto astTup = mod->stmt(1);
+  assert(astTup->isTupleDecl());
+  auto& qtTup = rr.byAst(astTup).type();
+  assert(qtTup.kind() == QualifiedType::CONST_REF);
+  assert(qtTup.type()->isTupleType());
+
+  auto m = resolveTypesOfVariables(context, program, { "a", "b" });
+  assert(!guard.realizeErrors());
+  assert(m["a"].kind() == QualifiedType::CONST_REF);
+  assert(m["a"].type()->isIntType());
+  assert(m["b"].kind() == QualifiedType::CONST_REF);
+  assert(m["b"].type()->isIntType());
+}
+
 int main() {
```

Comment:
Could you add one more test that ensures that `=` is not resolved for records when they are assigned to `_`? You added that logic but I don't see it tested here.

Reply:
Good reminder, thanks!!!